services/audit/audit.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package audit
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"time"
+
+	"code.gitea.io/gitea/models"
+	asymkey_model "code.gitea.io/gitea/models/asymkey"
+	auth_model "code.gitea.io/gitea/models/auth"
+	git_model "code.gitea.io/gitea/models/git"
+	organization_model "code.gitea.io/gitea/models/organization"
+	repository_model "code.gitea.io/gitea/models/repo"
+	secret_model "code.gitea.io/gitea/models/secret"
+	user_model "code.gitea.io/gitea/models/user"
+	webhook_model "code.gitea.io/gitea/models/webhook"
+	"code.gitea.io/gitea/modules/graceful"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/queue"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util/rotatingfilewriter"
+)
+
+type TypeDescriptor struct {
+	Type         string `json:"type"`
+	PrimaryKey   any    `json:"primary_key"`
+	FriendlyName string `json:"friendly_name"`
+	Target       any    `json:"-"`
+}
+
+type Event struct {
+	Action  Action         `json:"action"`
+	Doer    TypeDescriptor `json:"doer"`
+	Scope   TypeDescriptor `json:"scope"`
+	Target  TypeDescriptor `json:"target"`
+	Message string         `json:"message"`
+	Time    time.Time      `json:"time"`
+}
+
+var (
+	appenders  = make([]Appender, 0, 5)
+	auditQueue queue.Queue
+)
+
+func Init() {
+	if !setting.Audit.Enabled {
+		return
+	}
+
+	for name, opts := range setting.Audit.AppenderOptions {
+		var a Appender
+		switch name {
+		case "log":
```

Comment:
Is it possible to use this interface to write logs to S3 (for immutable access)? For funding part of the bounty this is something Copia would like.

Reply:
Personally I would suggest to only use database to store audit logs.

IMO the audit should be in database , and should be displayable in web UI.

Reply:
@wxiaoguang exporting audit logs or connecting to logging aggregators is a common requirement for this capability.

Reply:
OK, an update here: after asking around I can confirm it's fine to log to STDOUT, with the provision that it's easy to identify logs that are for the audit trail. Is it possible or can this PR be modified to make it possible to tag each log with something identifying it as an audit log?

Reply:
@philip-peterson I see that it now can write logs to files with rotation. Maybe those files can be used to import into log aggregators?

Reply:
> @wxiaoguang exporting audit logs or connecting to logging aggregators is a common requirement for this capability.

But IMO that is too difficult for a small org to use. Without a complex logger/viewer setup, they couldn't view the audit log on UI?

Reply:
@wxiaoguang There is a notice appender already which writes the logs as admin notices.

Reply:
Sure, that's not a problem. I just did not want to blow this "base" PR with all possible appender types. Adding the database would require the model and some sort of UI. Currently this PR just contains the base logging.

Reply:
Could we drop the "notice database appender" at the moment? I worry about if we have it now, then we couldn't remove it in the future (people always worry about breaking....)

Reply:
> Could we drop the "notice database appender" at the moment? I worry about if we have it now, then we couldn't remove it in the future (people always worry about breaking....)

ping ~~~

Reply:
pong